server/playground.rs

```diff
@@ -0,0 +1,139 @@
+use hubcaps::gists;
+use iron::{status, IronResult, Request, Response};
+use router::Router;
+use serde::{Deserialize, Serialize};
+use serde_json;
+use std::collections::HashMap;
+use std::convert::TryFrom;
+use std::io::Read;
+use tokio::runtime::Runtime;
+
+use crate::errors::Error;
+use crate::json;
+use crate::middleware::GistsMiddleware;
+
+#[derive(Debug, Serialize, Deserialize)]
+struct Playground {
+    id: Option<String>,
+    messages: String,
+    variables: serde_json::Value,
+    setup: serde_json::Value,
+}
+
+pub fn get(req: &mut Request) -> IronResult<Response> {
+    let gists_middleware = req.extensions.get::<GistsMiddleware>().unwrap();
+    let gists = &gists_middleware.gists;
+    let params = req.extensions.get::<Router>().unwrap();
+    let id = params.find("id").expect("No route parameter called id");
+    let mut rt = match Runtime::new() {
+        Ok(rt) => rt,
+        Err(_) => return json::error(status::ServiceUnavailable, Error::Runtime),
+    };
+    let gist = match rt.block_on(gists.get(id)) {
+        Ok(gist) => gist,
+        Err(hubcaps::errors::Error(hubcaps::errors::ErrorKind::Fault { code, .. }, _))
+            if code == 404 =>
+        {
+            return json::error(status::NotFound, Error::NotFound)
+        }
+        Err(_) => return json::error(status::InternalServerError, Error::Fetching),
+    };
```

Comment:
```rust
let gist = rt.block_on(gists.get(id)).unwrap_or_else(|err| if let Err(...) = err {
  not found case
} else {
  internal error case
});
```

Reply:
This doesn't let me return `json::error` early. I could still use `unwrap_or_else` or `map_err` to handle the error returned by `gists.get()` first, followed by the error returned by `block_on`. I think find the flatter variant using a single `match` easier on the eyes, but I'll be happy to reconsider.